### PR TITLE
Remove embedding layer norm from clip text encoder and some fixes

### PR DESCRIPTION
**Pull request by ankitade**
Differential Revision: D38020819

### CODE REVIEW

File: torchmultimodal/modules/encoders/clip_text_encoder.py [hunk 1]
```diff
@@ -39,42 +47,49 @@ def __init__(
...
+        self.encoder = TransformerEncoder(encoder_layer, num_layers=layers)
 
         self.width = width
         self.context_length = context_length
 
         self.ln_final = nn.LayerNorm(width)
         self.projection = nn.Linear(width, embedding_dim, bias=False)
 
-        if use_clip_init:
+        if pretrained_url:
```
Comment (langong347) [inline torchmultimodal/modules/encoders/clip_text_encoder.py:72]: Why not having `load_module_from_url` outside of the module so that:
-  you don't need to pass in `self` as the first argument? 
-  `pretrained` is more of an option to construct a model rather than part of the model logic

For example, in [Omnivore](https://github.com/facebookresearch/multimodal/blob/e274207e5c091d681bcb790d561fca0764148549/torchmultimodal/models/omnivore.py#L148) and [VideoCLIP](https://github.com/facebookresearch/multimodal/blob/main/examples/mugen/retrieval/video_clip.py#L163) we had `load_model` at the builder level.

File: torchmultimodal/modules/encoders/clip_text_encoder.py [hunk 2]
```diff
@@ -39,42 +47,49 @@ def __init__(
...
+        )
+        encoder_layer = TransformerEncoderLayer(
+            d_model=width,
+            nhead=heads,
             dropout=0.0,
-            normalize_before=True,
+            activation=quick_gelu,
+            norm_first=True,
         )
+        self.encoder = TransformerEncoder(encoder_layer, num_layers=layers)
```
Comment (langong347) [inline torchmultimodal/modules/encoders/clip_text_encoder.py:64]: this is to utilize PT core's better transformer? I thought text team has or is going to adopt it?

File: torchmultimodal/modules/encoders/clip_text_encoder.py [hunk 3]
```diff
@@ -87,14 +102,21 @@ def initialize_parameters(self) -> None:
 
     def build_attention_mask(self) -> torch.Tensor:
         # To support torchscripting, we have to pass an int as fill_value
-        mask = torch.full((self.context_length, self.context_length), int(True)).triu(1)
-        return mask.to(device=None, dtype=torch.bool)
+        mask = torch.full(
+            (self.context_length, self.context_length), float("-inf")
```
Comment (langong347) [inline torchmultimodal/modules/encoders/clip_text_encoder.py:106]: Why "-inf" here? Originally the mask was filled with "True" (or integer 1 for torchscript):

```
def build_attention_mask(self):
        mask = torch.full((self.context_length, self.context_length), True).triu(1)
        return mask.to(device=None, dtype=torch.bool)
```

cc @ebsmothers @sophiazhi

File: test/modules/encoders/test_clip_text_encoder.py [hunk 4]
```diff
@@ -24,89 +26,69 @@ def build_encoder(
...
             return CLIPTextEncoder(
                 embedding_dim=embedding_dim,
                 use_clip_init=use_clip_init,
                 context_length=context_length,
+                heads=heads,
             )
 
         return build_encoder, build_text
 
-    def test_initialization(self, start):
```
Comment (langong347) [inline test/modules/encoders/test_clip_text_encoder.py:36]: Do you mean to remove this test or plan to modify it? This was originally proposed to validate the effect of enabling clip initialization.

cc @sophiazhi  @ebsmothers

Comment (ebsmothers) [reply test/modules/encoders/test_clip_text_encoder.py:36]: I chatted with @ankitade about this yesterday and seems this test is a bit flaky, making it difficult to make changes to other parts of CLIP code without triggering failures. Given that the specific initialization is not used except for the pretraining from the paper (which we are not doing), I think it's OK to remove.

File: torchmultimodal/utils/common.py [hunk 5]
```diff
@@ -96,18 +97,14 @@ def tensor_slice(x: Tensor, begin: List[int], size: List[int]) -> Tensor:
 
 
 def load_module_from_url(
-    model: torch.nn.Module, url: str, progress: bool = True
+    model: torch.nn.Module, url: str, strict: bool = True, progress: bool = True
```
Comment (ebsmothers) [inline torchmultimodal/utils/common.py:100]: Are you abandoning #184?

File: torchmultimodal/modules/encoders/clip_text_encoder.py (hunk 2 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/clip_text_encoder.py:64]: They have, but they are also wrapping it with some other stuff (for instance their class also includes the embeddings).

File: torchmultimodal/modules/encoders/clip_text_encoder.py (hunk 3 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/clip_text_encoder.py:106]: Previously TorchText's transformer encoder was calling into their own MHA [here](https://github.com/pytorch/text/blob/ed44dc88539b51785df03a276410cf71adfe6bfd/torchtext/models/roberta/modules.py#L135-L137). But with the change to transformer encoder they are no longer doing this. So I think we need to make this change regardless

File: torchmultimodal/modules/encoders/clip_text_encoder.py (hunk 2 above)
Comment (RdoubleA) [reply torchmultimodal/modules/encoders/clip_text_encoder.py:64]: This should ultimately be updated by our unified TransformerEncoder

File: torchmultimodal/modules/encoders/clip_text_encoder.py (hunk 1 above)
Comment (ankitade) [reply torchmultimodal/modules/encoders/clip_text_encoder.py]: this was more for user passing in any ckpt path (vs predefined ones). but i pushed the responsibility for such things to the user for now

File: torchmultimodal/modules/encoders/clip_text_encoder.py (hunk 2 above)
Comment (ankitade) [reply torchmultimodal/modules/encoders/clip_text_encoder.py:61]: the goal here is to get rid of the layer norm after embedding, using core's transformer is just a side effect

**Review by ebsmothers — APPROVED** [review]
